### PR TITLE
Create a Chunk from an IterableOnce

### PR DESCRIPTION
**Pull request by peterneyens**
The implementation is basically the one from `iterable` and `chain`.

I replaced the implementations in `iterable` and `chain` with a call to `iterableOnce`. This does add an `Iterator#isEmpty` call.

### CODE REVIEW

File: core/shared/src/main/scala/fs2/Chunk.scala [hunk 1]
```diff
@@ -596,18 +596,24 @@ object Chunk
...
     })
 
+  /** Creates a chunk from a `scala.collection.IterableOnce`. */
+  def iterableOnce[O](i: collection.IterableOnce[O]): Chunk[O] = {
+    val itr = i.iterator
+    if (itr.isEmpty) empty
+    else {
+      val head = itr.next()
+      if (itr.hasNext) {
+        val bldr = collection.mutable.Buffer.newBuilder[O]
```
Comment (nikiforo) [inline core/shared/src/main/scala/fs2/Chunk.scala:609]: According to this
https://github.com/typelevel/fs2/pull/2720/files#diff-e8fdf3ccd3d2479432355b3abf34a16cd548373afeddeaf231b8d8a1147b1d48R115
ArrayBuilder shows x2 performance compared to mutable.Buffer, shouldn't we change a type of builder used here?

File: core/shared/src/main/scala/fs2/Chunk.scala [hunk 2]
```diff
@@ -596,18 +596,24 @@ object Chunk
       case ix: GIndexedSeq[O] => indexedSeq(ix)
       case _ =>
         if (i.isEmpty) empty
```
Comment (nikiforo) [inline core/shared/src/main/scala/fs2/Chunk.scala:598]: emptiness gets checked in `iterableOnce`, can't we remove this check from here and `L626`?

File: core/shared/src/main/scala/fs2/Chunk.scala (hunk 1 above)
Comment (peterneyens) [reply core/shared/src/main/scala/fs2/Chunk.scala:609]: In those cases we have the (expected) size, we don't in this case.

**Review by diesalbla — APPROVED** [review]

**Comment by peterneyens** [comment]
I already forgot `IterableOnce` wasn't a thing in 2.12.

File: core/shared/src/main/scala/fs2/Chunk.scala (hunk 1 above)
Comment (mpilquist) [reply core/shared/src/main/scala/fs2/Chunk.scala:608]: ArrayBuilder should still beat mutable buffer though, even without a size hint.

File: core/shared/src/main/scala/fs2/Chunk.scala (hunk 2 above)
Comment (nikiforo) [reply core/shared/src/main/scala/fs2/Chunk.scala:598]: Now the implementation has changed, this comment is outdated

File: core/shared/src/main/scala/fs2/Chunk.scala (hunk 1 above)
Comment (peterneyens) [reply core/shared/src/main/scala/fs2/Chunk.scala:608]: Using `ArrayBuilder` requires us to add a `ClassTag` constraint, I don't think that is something that we want?

That would mean that we couldn't use it in `iterable` and `chain`.

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/Chunk.scala:608]: Merge the main branch in to this PR and then use `makeClassTag[Any]`. We'll have to cast the resulting chunk back to a `Chunk[A]` just like we do in methods like `filter`.

**Comment by mpilquist** [comment]
Thanks!

**Pull request merged**